README.md
# UShoard

The goal of UShoard is to show Data of Sales Volume of US Treasury Bills and Bonds

## Example

This is a basic example which shows  how to aquire US Treasure Bill and Bonds time indexed by 'issue date'

```r
#' ## issue dates
#' treasuryTRISS10YEARNOTE <- getSymbols("TRISS10YEARNOTE", src = "treasury", auto.assign = FALSE)
#' head(treasuryTRISS10YEARNOTE)
#'
```
This is a basic example which shows  how to aquire US Treasure Bill and Bonds time indexed by 'maturity date'

```r
#' ## maturity dates
#' treasuryTRMTR10YEARNOTE <- getSymbols("TRMTR10YEARNOTE", src = "treasury", auto.assign = FALSE)
#' head(treasuryTRMTR10YEARNOTE)
#' tail(treasuryTRMTR10YEARNOTE)
```

This is an example of the package in action

```r
#' 
#' > treasuryTRISS10YEARNOTE <- getSymbols("TRISS10YEARNOTE", src = "treasury", auto.assign = FALSE)
#' 
#' > treasuryTRISS10YEARNOTE[, "TRISS10YEARNOTE.TOTAL_ISSUE"]["2000/2001"]
#'            TRISS10YEARNOTE.TOTAL_ISSUE
#' 2000-02-15                   12277.576
#' 2000-05-15                   11078.133
#' 2000-08-15                   12360.050
#' 2000-11-15                   10077.544
#' 2001-02-15                   11975.972
#' 2001-05-15                   11460.357
#' 2001-08-15                   12045.667
#' 2001-10-05                    6000.038 # DOWNTREND TOO LATE
#' 2001-11-15                    8589.611
#' 
#' > treasuryTRISS10YEARNOTE[, "TRISS10YEARNOTE.TOTAL_ISSUE"]["2007/2008"]
#'            TRISS10YEARNOTE.TOTAL_ISSUE
#' 2007-02-15                       14193
#' 2007-03-15                        8000
#' 2007-05-15                       17586
#' 2007-06-15                        8000
#' 2007-08-15                       20000
#' 2007-09-17                        8000
#' 2007-11-15                       19675
#' 2007-12-17                        8000
#' 2008-02-15                       19205
#' 2008-03-17                       10000
#' 2008-05-15                       23078
#' 2008-06-16                       11000 # TOO LATE SLIGHT UPSWING BEGINS
#' 2008-08-15                       23339
#' 2008-09-15                       13464
#' 2008-10-15                       10000
#' 2008-10-15                       10000
#' 2008-10-15                       10000
#' 2008-10-15                       10000
#' 2008-11-17                       23549
#' 2008-12-15                       16479
#' 
#' > treasuryTRISS2YEARNOTE <- getSymbols("TRISS2YEARNOTE", src = "treasury", auto.assign = FALSE)
#' 
#' > treasuryTRISS2YEARNOTE[, "TRISS2YEARNOTE.TOTAL_ISSUE"]["2000/2001"]
#'            TRISS2YEARNOTE.TOTAL_ISSUE
#' 2000-01-31                   19381.25
#' 2000-02-29                   16563.38
#' 2000-03-31                   17237.94
#' 2000-05-01                   17390.90
#' 2000-05-31                   14871.82
#' 2000-06-30                   14320.61
#' 2000-07-31                   15057.90
#' 2000-08-31                   15072.21
#' 2000-10-02                   15144.33
#' 2000-10-31                   14856.61
#' 2000-11-30                   15058.53
#' 2001-01-02                   14822.03
#' 2001-01-31                   15452.60 # NO REAL TREND
#' 2001-02-28                   14685.10
#' 2001-04-02                   14674.85
#' 2001-04-30                   13338.53
#' 2001-05-31                   13331.94
#' 2001-07-02                   14671.07
#' 2001-07-31                   16003.27 # UPSWING BEGINS (TOO LATE)
#' 2001-08-31                   18665.04
#' 2001-10-01                   22675.48
#' 2001-10-31                   25147.96
#' 2001-11-30                   26170.53
#' 2001-12-31                   29666.99
#' 
#' > treasuryTRISS2YEARNOTE[, "TRISS2YEARNOTE.TOTAL_ISSUE"]["2007/2008"]
#'            TRISS2YEARNOTE.TOTAL_ISSUE
#' 2007-01-02                      24817
#' 2007-01-31                      22976
#' 2007-02-28                      23038
#' 2007-04-02                      22645
#' 2007-04-30                      22778
#' 2007-05-31                      22487
#' 2007-07-02                      22021
#' 2007-07-31                      21888
#' 2007-08-31                      22282
#' 2007-10-01                      21969
#' 2007-10-31                      23866
#' 2007-11-30                      24544 # NOTICABLE UPSWING IN ISSUES
#' 2007-12-31                      26497
#' 2008-01-31                      26925
#' 2008-02-29                      28900
#' 2008-03-31                      31992
#' 2008-04-30                      32753
#' 2008-06-02                      32410
#' 2008-06-30                      33219
#' 2008-07-31                      34422
#' 2008-09-02                      34670
#' 2008-09-30                      37148
#' 2008-10-31                      35507
#' 2008-12-01                      38520
#' 2008-12-31                      40777
#' 
#' > treasuryTRISS13WEEKBILL <- getSymbols("TRISS13WEEKBILL", src = "treasury", auto.assign = FALSE)
#' 
#' > treasuryTRISS13WEEKBILL[, "TRISS13WEEKBILL.TOTAL_ISSUE"]["2000/2001"]
#'            TRISS13WEEKBILL.TOTAL_ISSUE # DID NOT START TRACKING UNTIL LATE 2001
#' 2001-08-02                    18401.50
#' 2001-08-09                    19657.91
#' 2001-08-16                    19161.88
#' 2001-08-23                    19096.12
#' 2001-08-30                    21848.33
#' 2001-09-06                    18488.62
#' 2001-09-13                    18403.86
#' 2001-09-20                    18324.61
#' 2001-09-27                    18538.18
#' 2001-10-04                    17315.95
#' 2001-10-11                    16123.83
#' 2001-10-18                    17318.82
#' 2001-10-25                    18487.95
#' 2001-11-01                    20386.58
#' 2001-11-08                    21112.85
#' 2001-11-15                    20840.48
#' 2001-11-23                    20866.16
#' 2001-11-29                    22916.21
#' 2001-12-06                    20616.11
#' 2001-12-13                    18373.31
#' 2001-12-20                    18278.22
#' 2001-12-27                    18599.52
#' 
#' > treasuryTRISS13WEEKBILL[, "TRISS13WEEKBILL.TOTAL_ISSUE"]["2007/2008"]
#'            TRISS13WEEKBILL.TOTAL_ISSUE
#' 2007-01-04                       24081
#' 2007-01-11                       23893
#' 2007-01-18                       24049
#' 2007-01-25                       24218
#' 2007-02-01                       24156
#' 2007-02-08                       26750
#' 2007-02-15                       26777
#' 2007-02-22                       28166
#' 2007-03-01                       29273
#' 2007-03-08                       29162
#' 2007-03-15                       29104
#' 2007-03-22                       27762
#' 2007-03-29                       26599
#' 2007-04-05                       24088
#' 2007-04-12                       21602
#' 2007-04-19                       20460
#' 2007-04-26                       19141
#' 2007-05-03                       19134
#' 2007-05-10                       19651
#' 2007-05-17                       20939
#' 2007-05-24                       22607
#' 2007-05-31                       22995
#' 2007-06-07                       23442
#' 2007-06-14                       22203
#' 2007-06-21                       21968
#' 2007-06-28                       21876
#' 2007-07-05                       22746
#' 2007-07-12                       22502
#' 2007-07-19                       23253 
#' 2007-07-26                       24237
#' 2007-08-02                       27824 # UPSWING IN SALES
#' 2007-08-09                       29149
#' 2007-08-16                       29270
#' 2007-08-23                       26243
#' 2007-08-30                       28434
#' 2007-09-06                       30068
#' 2007-09-13                       27674
#' 2007-09-20                       24413
#' 2007-09-27                       21589 # DOWNSWING IN SALES
#' 2007-10-04                       21903
#' 2007-10-11                       21856
#' 2007-10-18                       21857
#' 2007-10-25                       23400
#' 2007-11-01                       25679 # UPSWING IN SALES
#' 2007-11-08                       27421
#' 2007-11-15                       27480
#' 2007-11-23                       26561
#' 2007-11-29                       29258
#' 2007-12-06                       25998
#' 2007-12-13                       25413
#' 2007-12-20                       21001 # DOWNSWING IN SALES
#' 2007-12-27                       20003
#' 2008-01-03                       20001
#' 2008-01-10                       24309
#' 2008-01-17                       21530
#' 2008-01-24                       22412
#' 2008-01-31                       24390
#' 2008-02-07                       29082 # UPSWING IN SALES
#' 2008-02-14                       29360
#' 2008-02-21                       27348
#' 2008-02-28                       27920
#' 2008-03-06                       27764
#' 2008-03-13                       25881
#' 2008-03-20                       24000 
#' 2008-03-27                       23996
#' 2008-04-03                       24000
#' 2008-04-10                       24003
#' 2008-04-17                       24381
#' 2008-04-24                       21383
#' 2008-05-01                       20007 # BOTTOM POINT OF SALES
#' 2008-05-08                       22005
#' 2008-05-15                       24008
#' 2008-05-22                       25006
#' 2008-05-29                       25000
#' 2008-06-05                       24002
#' 2008-06-12                       24006
#' 2008-06-19                       23006
#' 2008-06-26                       22011
#' 2008-07-03                       23000
#' 2008-07-10                       24003
#' 2008-07-17                       24020
#' 2008-07-24                       24022
#' 2008-07-31                       24016
#' 2008-08-07                       24998
#' 2008-08-14                       26008
#' 2008-08-21                       28013
#' 2008-08-28                       28006
#' 2008-09-04                       27017
#' 2008-09-11                       26006
#' 2008-09-18                       28017 # TOP POINT OF SALES ( REMAINS HIGH )
#' 2008-09-25                       27018
#' 2008-10-02                       26021
#' 2008-10-09                       26015
#' 2008-10-16                       25012
#' 2008-10-23                       25013
#' 2008-10-30                       25003
#' 2008-11-06                       26997
#' 2008-11-13                       26998
#' 2008-11-20                       26997
#' 2008-11-28                       28013
#' 2008-12-04                       28000
#' 2008-12-11                       26998
#' 2008-12-18                       27007
#' 2008-12-26                       27008 # HIGHNESS IS RETAINED
#' 
#' # lately
#' 
#' > treasuryTRISS13WEEKBILL[, "TRISS13WEEKBILL.TOTAL_ISSUE"]["2018"]
#'            TRISS13WEEKBILL.TOTAL_ISSUE
#' 2018-01-04                    48.00076
#' 2018-01-11                    48.00043
#' 2018-01-18                    47.99864
#' 2018-01-25                    48.00002
#' 2018-02-01                    48.00687
#' 2018-02-08                    48.00233
#' 2018-02-15                    48.00154
#' 2018-02-22                    50.99863
#' 2018-03-01                    51.00135
#' 2018-03-08                    50.99683
#' 2018-03-15                    50.99852
#' 2018-03-22                    51.00521
#' 2018-03-29                    51.00016
#' 2018-04-05                    48.00078
#' 2018-04-12                    48.00028
#' 2018-04-19                    48.01236
#' 2018-04-26                    48.00147
#' 2018-05-03                    48.00222
#' 2018-05-10                    48.00294
#' 2018-05-17                    48.00126
#' 2018-05-24                    47.99824
#' 2018-05-31                    48.00532
#' 2018-06-07                    47.99921
#' 2018-06-14                    48.00599
#' 2018-06-21                    48.00364
#' 2018-06-28                    47.99992
#' 2018-07-05                    47.95970
#' 2018-07-12                    48.00011
#' 2018-07-19                    51.00118
#' 2018-07-26                    51.00867
#' 2018-08-02                    51.00326
#' 2018-08-09                    51.00147
#' 2018-08-16                    51.00094
#' 2018-08-23                    51.01022
#' 2018-08-30                    51.00416
#' >
#' 
```


NEWS.md
# UShoard 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.





tests/testthat.R
library(testthat)
library(UShoard)

test_check("UShoard")


tests/testthat/test-all.R


context("working")

# test_that("aquire data", {
#   expect_equal(
#       rawData(normalizePath(paste0(path.package("UShoard"),"/inst/September 10_2018 IC Coupons.xls"), winslash = "/", mustWork = TRUE)
#         , if(.Platform$OS.type == "windows") { "NUL" } else { "/dev/null" }
#       )
#     , NULL)
# })

test_that("format data", {

  expect_equal(  formatTreasuriesAllotmentDateCols(
                   data.frame(alpha=1:2), "alpha"
                 )
               , structure(list(alpha = structure(c(1, 2), class = "Date")), row.names = c(NA, -2L), class = "data.frame")
               )

  # SHOULD need A TEST here
  expect_equal("formatReportAllotment", "formatReportAllotment")

  expect_equal( reorgRawTreasuriesAllotment(
                  data.frame(security_term = c("BOND", "NOTE", "YR")
                , security_type = rep("",3)
                , issue_date = rep(NA,3)
                , stringsAsFactors = FALSE)
              )
              , structure(  list(issue_date = c(NA, NA, NA)
                          , security = c("Bond-", "Note-", "Year-"))
                          , class = "data.frame", row.names = c(NA, -3L)
                         )
              )

  expect_equal( reorgRawTreasuriesAllotment(
                  data.frame(security_term = c("CASH", "WK", "YR")
                , issue_date = rep(NA,3)
                , stringsAsFactors = FALSE)
              )
              , structure(  list(issue_date = c(NA, NA, NA)
                          , security = c("Cash Bill", "Week Bill", "Year Bill"))
                          , class = "data.frame", row.names = c(NA, -3L)
                         )
              )

  expect_equal("recentRawTreasuriesAllotment",      "recentRawTreasuriesAllotment")
  expect_equal("recentRawTreasuriesAllotments"    , "recentRawTreasuriesAllotments")

  expect_equal("historicalRawTreasuriesAllotment", "historicalRawTreasuriesAllotment")
  expect_equal("historicalRawTreasuriesAllotments", "historicalRawTreasuriesAllotments")

  expect_equal("rawTreasuriesAllotments", "rawTreasuriesAllotments")
})


R/UShoard.R
#' UShoard: Data of Sales Volume of US Treasury Bills and Bonds
#'
#' @section Source:
#'
#' Investor Class Auction Allotments
#' U.S. DEPARTMENT OF THE TREASURY
#' https://home.treasury.gov/data/investor-class-auction-allotments
#'
#' RECENT DATA
#' Coupon Auctions - Data from October 2009-present
#' This table provides investor class allotments for marketable Treasury coupon auctions.
#' Data are ordered by issue date.
#' This table was previously published as the PDO-4 table in the quarterly Treasury Bulletin.
#' Release Dates:  Release Dates:  September 22nd, October 7th, October 22nd
#'
#' Bill Auctions - Data from October 2009-present
#' This table provides investor class allotments for marketable Treasury bill auctions.
#' Data are ordered by issue date.
#' Release Dates: October 7th
#'
#' HISTORICAL DATA
#' Coupon Auctions - Data from January 2000-September 2009
#' This table provides investor class allotments for marketable Treasury coupon auctions.
#' Data are ordered by issue date.
#' This table was previously published as the PDO-4 table in the quarterly Treasury Bulletin.
#'
#' Bill Auctions - Data from August 2001-September 2009
#' This table provides investor class allotments for marketable Treasury bill auctions.
#' Data are ordered by issue date.
#'
#' @section Treasures Securities:
#'
#' Of the following treasury securities,
#' included but not limited to
#'
#' 4, 13, 26, and 52 Week Bills,
#' 2-Year FRN,
#' 2-Year Note,
#' 5-Year TIPS Note,
#' 5-Year Note,
#' 7-Year Note,
#' 3-Year Note,
#' 10-Year Note,
#' and
#' 30-Year Bond
#'
#' @section Provides:
#'
#' issue_date,
#' security -
#'     term or type condensed to security,
#' rate -
#'     coupon rate, or spread, or auction high rate,
#' cusip,
#' maturity_date,
#' total_issue
#'
#' Also total_issue is further detailed.
#'
#' soma_issue,
#' institutions_issue,
#' individuals_issue,
#' dealers_issue,
#' pensions_issue,
#' investment_funds_issue,
#' foreign_issue,
#' and
#' other_issue
#'
#' @docType package
#' @name UShoard
NULL


R/main.R

#' sets the enviroment
#'
#' space-saver - meant to be used at the beginning of a function
#'
#' @param init list with names entries of alernate/other options
#' @param envir where to return the options
#' @return environment is set
#' @examples
#' \dontrun{
#' # > options(max.print=88888L)
#' # initEnv()
#' # > getOption("max.print")
#' # [1] 99999
#' }
initEnv <- function(init = NULL, envir = parent.frame()) {
  tryCatchLog::tryCatchLog({
  require(data.table)
  require(xts)
  require(quantmod)

  eval(parse(text="assign(\"env\", environment())"), envir = parent.frame())

  assign("%>%", magrittr::`%>%`, envir = envir)

  assign("str_detect", stringr::str_detect, envir = envir)
  assign("str_c",           stringr::str_c, envir = envir)
  assign("str_replace",     stringr::str_replace, envir = envir)
  assign("str_replace_all", stringr::str_replace_all, envir = envir)

  assign("MoveFront", DataCombine::MoveFront, envir = envir)

  assign("as.Date", zoo::as.Date, envir = envir)

  assign("select",      dplyr::select, envir = envir)
  assign("mutate",      dplyr::mutate, envir = envir)
  assign("filter",      dplyr::filter, envir = envir)

  ops <- options()

  # tryCatchLog: what level to activate
  futile.logger::flog.threshold(futile.logger::ERROR)

  options(warn=2L)
  assign("ops", ops, envir = envir)

  invisible()

})}


#' unsets the enviroment
#'
#' space-saver - meant to be used at the beginning of a function
#'
#' @return environment is un-set
#' @examples
#' \dontrun{
#' # > uninitEnv()
#' # getOption("digits")
#' # [1] 5
#' }
uninitEnv <- function() {
  tryCatchLog::tryCatchLog({

  options(get("ops", envir = parent.frame()))
  invisible()

})}


#' Get data from the internet/filesystem
#'
#' Aquire data.
#'
#' @param url internet url or filesystem path
#' @param newLocalFileName current area location to save
#' @examples
#' \dontrun{
#'
#' # recent coupon auctions
#' # Coupon Auctions – Data from October 2009-present?
#' # url <- "https://home.treasury.gov/system/files/291/September%2010_2018%20IC%20Coupons.xls"
#' rawData(url = url, newLocalFileName = "recentCoupons.xls")
#' df <- as.data.frame(
#'       suppressWarnings(readxl::read_excel("recentCoupons.xls"
#'         , col_types = "guess", skip  = 4, guess_max = 1)
#'       ))
#' col_names <- c("issue_date", "security", "coupon_rate"
#'              , "cusip", "maturity_date", "total_issue"
#'              , "SOMA_issue", "institutions_issue", "individuals_issue"
#'              , "dealers_issue", "pensions_issue"
#'              , "investment_funds_issue", "foreign_issue","other_issue"
#'              )
#' colnames(df) <- col_names
#' df$issue_date    <- zoo::as.Date(df$issue_date)
#' df$maturity_date <- zoo::as.Date(df$maturity_date)
#' str(df, vec.len = 3)
#' # 'data.frame':	803 obs. of  14 variables:
#' #  $ issue_date            : Date, format: "2018-08-31" "2018-08-31" "2018-08-31" ...
#' #  $ security              : chr  "2-Year FRN" "2-Year Note" "5-Year TIPS Note" ...
#' #  $ rate                  : num  0.043 2.625 0.625 2.75 ...
#' #  $ cusip                 : chr  "912828Y53" "9128284Y3" "9128284H0" ...
#' #  $ maturity_date         : Date, format: "2020-07-31" "2020-08-31" "2023-04-15" ...
#' #  $ total_issue           : num  18.2 38.5 15 39.6 ...
#' #  $ soma_issue            : num  1.198 2.537 0.987 2.608 ...
#' #  $ institutions_issue    : num  0 0.001 0 0.07 0.07 0.00051 0 0 ...
#' #  $ individuals_issue     : num  0.0203 0.393 0.0351 0.056 ...
#' #  $ dealers_issue         : num  8.9 15.7 2.73 9.8 ...
#' #  $ pensions_issue        : num  0.217 0.025 0 0 ...
#' #  $ investment_funds_issue: num  3.72 15.37 10.52 23.49 ...
#' #  $ foreign_issue         : num  0.49 4.426 0.715 3.134 ...
#' #  $ other_issue           : num  3.65 0.08 0 0.444 ...
#' }
#' @export
rawData <- function(url = NULL, newLocalFileName = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  if(str_detect(url, "^http")) {
     # url <- URLdecode(URL = url)
    download.file(url = url, destfile = newLocalFileName, mode = "wb")
  } else {
     file.copy(from = url, to = newLocalFileName, overwrite = TRUE, copy.mode = TRUE, copy.date = TRUE)
  }
  invisible()

})}


#' Get one recent Investor Class Auction Allotment
#'
#' @param treasClassUrl url of one single recent allotment
#' @examples
#' \dontrun{
#' recentRaw <- recentRawTreasuriesAllotment()
#' recentRaw
#' }
#' @export
recentRawTreasuriesAllotment <- function(treasClassUrl = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv();unlink(tmpf)})

  # recent coupon auctions
  # Coupon Auctions – Data from October 2009-present?
  if(is.null(treasClassUrl)) treasClassUrl <- "https://home.treasury.gov/system/files/291/September%2010_2018%20IC%20Coupons.xls"
  tmpf <- tempfile()
  rawData(url = treasClassUrl, newLocalFileName = tmpf)

  df <- as.data.frame(
        suppressWarnings(readxl::read_excel(tmpf, col_types = "guess", skip  = 4, guess_max = 1)
        ))                                # rate: action rate, coupon rate, or spread
  col_names <- c("issue_date", "security", "rate", "cusip"
               , "maturity_date", "total_issue", "SOMA_issue", "institutions_issue"
               , "individuals_issue", "dealers_issue", "pensions_issue"
               , "investment_funds_issue", "foreign_issue","other_issue"
               )
  colnames(df)     <- col_names
  df <- formatTreasuriesAllotmentDateCols(df, cols = c("issue_date","maturity_date"))

  df

})}


#' Get many recent Investor Class Auction Allotments
#'
#' @param treasList list() containting urls of recent allotments
#' @examples
#' \dontrun{
#' recents <- recentRawTreasuriesAllotment()
#' recents
#' }
#' @export
recentRawTreasuriesAllotments <- function(treasList = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  if(is.null(treasList)) treasList <- list("https://home.treasury.gov/system/files/291/September%2010_2018%20IC%20Coupons.xls"
                                          ,"https://home.treasury.gov/system/files/291/September%2010_2018%20IC%20Bills_1.xls")
  treasAllotments <- data.frame()
  for (treasList_i in treasList) {
     treasAllotments <- rbind(treasAllotments, recentRawTreasuriesAllotment(treasClassUrl = treasList_i))
  }

  treasAllotments

})}


#' Get one historical Investor Class Auction Allotment
#'
#' @param treasClassUrl url of one historical allotment
#' @examples
#' \dontrun{
#'  historical <- historicalRawTreasuriesAllotment()
#'  historical
#' }
#' @export
historicalRawTreasuriesAllotment <- function(treasClassUrl = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  # recent coupon auctions
  # Coupon Auctions – Data from old to just before October?
  if(is.null(treasClassUrl)) treasClassUrl <- "https://www.treasury.gov/resource-center/data-chart-center/Documents/Website%20PDO-4-A-Coupons%20Jan%202000-Sep%202009.xls"
  tempfile <- tempfile()
  rawData(url = treasClassUrl, newLocalFileName = tempfile)

  df <- as.data.frame(
        suppressWarnings(readxl::read_excel(tempfile, col_types = "guess", skip  = 4, guess_max = 1)
        ))

  if(grepl("Coupons", treasClassUrl))  { security_type = "security_type" }
  if(grepl("Bills",    treasClassUrl)) { security_type = NULL } # rate: action rate, coupon rate, or spread
  col_names <- c("issue_date",  security_type, "security_term", "rate", "cusip"
               , "maturity_date", "total_issue", "SOMA_issue", "institutions_issue"
               , "individuals_issue", "dealers_issue", "pensions_issue"
               , "investment_funds_issue", "foreign_issue","other_issue"
               )
  colnames(df)     <- col_names

  df <- formatTreasuriesAllotmentDateCols(df, cols = c("issue_date","maturity_date"))

  df

})}



#' format Investor Class Auction Allotments date-ish columns
#'
#' change data columns to be of  R class Date
#'
#' @param treasuriesAllot Auction Allotments
#' @param cols columns in a character vector, to change to type Date
#' @examples
#' \dontrun{
#' formatted <- formatTreasuriesAllotmentDateCols(
#'                 data.frame(alpha=1:2), "alpha"
#'               )
#' formatted
#' }
#' @export
formatTreasuriesAllotmentDateCols <- function(treasuriesAllot = NULL, cols = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  if(!length(cols)) stop("To formatTreasuriesAllotmentDateCol on must pass cols (column names)")

  for(cols_i in cols) {
    treasuriesAllot[ , cols_i] <- as.Date(treasuriesAllot[ , cols_i] )
  }

  treasuriesAllot

})}



#' reorganize Investor Class Auction Allotments
#'
#' for 'old' Class Auction Allotments, change 'security_type' and/or 'security_term'
#' to be just (condensed to one column) 'security' with proper naming
#' compatible with the 'new'(er) Class Auction Allotments
#'
#' @param treasAllotment treasury allotmetn data.frame to be reorganized
#' @examples
#' \dontrun{
#' reorgedTreasury1 <-  reorgRawTreasuriesAllotment(
#'                data.frame(security_term = c("BOND", "NOTE", "YR")
#'              , security_type = rep("",3)
#'              , issue_date = rep(NA,3)
#'              , stringsAsFactors = FALSE)
#'            )
#' reorgedTreasury1
#'
#' reorgedTreasury2 <-  reorgRawTreasuriesAllotment(
#'                data.frame(security_term = c("CASH", "WK", "YR")
#'              , issue_date = rep(NA,3)
#'              , stringsAsFactors = FALSE)
#'            )
#' reorgedTreasury2
#' }
#' @export
reorgRawTreasuriesAllotment <- function(treasAllotment = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

    if(any(colnames(treasAllotment) %in% c("security_type", "security_term"))) {
      if(all(c("security_type", "security_term") %in% colnames(treasAllotment))) {
         treasAllotment
         with( treasAllotment, {
           security <- str_c(security_term,  " ", security_type)
           str_replace(security, "NOTE", "Note") %>%
             str_replace("BOND", "Bond")            %>%
               str_replace("YR",  "Year")             %>%
                 str_replace_all("-", " ")              %>%
                   str_replace(" ", "-")                  -> # first space only
                     security
        }) -> security
      } else if(all(c("security_term") %in% colnames(treasAllotment))) {
         with( treasAllotment, {
           str_replace(security_term, "CASH", "Cash") %>%
             str_replace("WK", "Week")                  %>%
               str_replace("YR",  "Year")                 %>%
                str_replace(" ", "-")                        %>% # first space only
                  str_replace("$", " Bill")                    ->
                     security
        }) -> security
      }

    }

    within(treasAllotment, {
        suppressWarnings(rm("security_type", "security_term"))
        assign("security", security)
    }) %>%
    MoveFront(c("issue_date", "security")) -> treasAllotment

     treasAllotment

})}



#' Get historical Investor Class Auction Allotments
#'
#' @param treasList list() containting urls of historical allotments
#' @param reorgRawTreasuriesAllotmentFn
#' @examples
#' \dontrun{
#'  rawHistoricals <- historicalRawTreasuriesAllotments()
#'  rawHistoricals
#' }
#' @export
historicalRawTreasuriesAllotments <- function(treasList = NULL, reorgRawTreasuriesAllotmentFn = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  if(is.null(treasList)) treasList <- list("https://www.treasury.gov/resource-center/data-chart-center/Documents/Website%20PDO-4-A-Coupons%20Jan%202000-Sep%202009.xls"
                                          ,"https://www.treasury.gov/resource-center/data-chart-center/Documents/Website%20IC%20allotments---Bills--Aug%202001-Sep%202009.xls")
  treasAllotments <- data.frame()
  for (treasList_i in treasList) {

    if(is.null(reorgRawTreasuriesAllotmentFn)) reorgRawTreasuriesAllotmentFn <- reorgRawTreasuriesAllotment
    historicalRawTreasuriesAllotment(treasClassUrl = treasList_i) %>%
       reorgRawTreasuriesAllotmentFn %>%
         { rbind(treasAllotments,. , stringsAsFactors = FALSE)} ->
          treasAllotments

  }

  treasAllotments

})}


#' Get Investor Class Auction Allotments
#'
#' Get both historical and recent Investor Class Auction Allotments
#'
#' @param treasuriesAllotments character vector of function names
#' @examples
#' \dontrun{
#' rawTreasuries <- rawTreasuriesAllotments()
#' rawTreasuries
#' }
#' @export
rawTreasuriesAllotments <- function(treasuriesAllotments = NULL) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  if(is.null(treasuriesAllotments)) treasuriesAllotments <- c("historicalRawTreasuriesAllotments"
                                                            , "recentRawTreasuriesAllotments")

  allotments <- data.frame()
  for (treasuriesAllotments_i in treasuriesAllotments) {
    rawTreasuriesAllotments_i <- do.call(treasuriesAllotments_i,list())

    allotments <- rbind(allotments, rawTreasuriesAllotments_i, stringsAsFactors = FALSE)

  }

  allotments

})}


#' Quantmod internal conversion
#'
#'
#' @param fr data.frame of information: left most column is the 'time' column
#' @param return.class call of the object sent back to the user
#' @examples
#' \dontrun{
#' quantmod___convert.time.series <- "rawTreasuriesAllotments()""
#' }
quantmod___convert.time.series <- function (fr, return.class) {
    tryCatchLog::tryCatchLog({
    initEnv();on.exit({uninitEnv()})
    if ("quantmod.OHLC" %in% return.class) {
        class(fr) <- c("quantmod.OHLC", "zoo")
        return(fr)
    }
    else if ("xts" %in% return.class) {
        return(fr)
    }
    if ("zoo" %in% return.class) {
        return(as.zoo(fr))
    }
    else if ("ts" %in% return.class) {
        fr <- as.ts(fr)
        return(fr)
    }
    else if ("data.frame" %in% return.class) {
        fr <- as.data.frame(fr)
        return(fr)
    }
    else if ("matrix" %in% return.class) {
        fr <- as.data.frame(fr)
        return(fr)
    }
    else if ("timeSeries" %in% return.class) {
        if (requireNamespace("timeSeries", quietly = TRUE)) {
            fr <- timeSeries::timeSeries(coredata(fr), charvec = as.character(index(fr)))
            return(fr)
        }
        else {
            warning(paste("'timeSeries' from package 'timeSeries' could not be loaded:",
                " 'xts' class returned"))
        }
    }
})}



#' format df in preparation for dplyr (future)selection
#'
#' @param df dplyr compatabile input type
#' @param AllotCat alotment category point of view, either issue "ISS" or maturity "MTR"
#' @param AllotmentCol 'issue_date" xor 'maturity_date"
#' @param unAllotmentCol 'maturity_date' xor 'issue_date"
#' @examples
#' \dontrun{

#' NEED one
#'
#' }
formatReportAllotment <- function(df, AllotmentCat, AllotmentCol, unAllotmentCol ) {
  tryCatchLog::tryCatchLog({
  initEnv();on.exit({uninitEnv()})

  select(df, - "cusip", - unAllotmentCol) %>%
    mutate( security = str_c(AllotmentCat, security) ) %>%
      MoveFront(c("security", AllotmentCol, "rate")) -> df

  mutate(df, security = toupper(security)
         , security = str_c("TR", security)
         , security = str_replace_all(security, "[ ]+", "")
         , security = str_replace_all(security,   "-+", "")
      ) -> df
  df

})}



#' Aquire Bills and Bonds allotment data from the US treasury
#'
#' @param Symbols One or more treasury 'issue' or 'maturity' bills/bonds symbols
#' @param env possibly returned data is loaded into this environment
#' @param return.class call of the object sent back to the user
#' @examples
#' \dontrun{
#' library(quantmod)
#'
#' ## issue dates
#' treasuryTRISS10YEARNOTE <- getSymbols("TRISS10YEARNOTE", src = "treasury", auto.assign = FALSE)
#' head(treasuryTRISS10YEARNOTE)
#'
#' ## maturity dates
#' treasuryTRMTR10YEARNOTE <- getSymbols("TRMTR10YEARNOTE", src = "treasury", auto.assign = FALSE)
#' head(treasuryTRMTR10YEARNOTE)
#'
#' }
#' @export
getSymbols.treasury <- function (Symbols, env, return.class = "xts", ...) {
    tryCatchLog::tryCatchLog({
    initEnv();on.exit({uninitEnv()})

    importDefaults("getSymbols.treasury")
    this.env <- environment()
    for (var in names(list(...))) {
        assign(var, list(...)[[var]], this.env)
    }
    if (!hasArg("verbose"))
        verbose <- FALSE
    if (!hasArg("auto.assign"))
        auto.assign <- TRUE

    treasuries <- rawTreasuriesAllotments()
    issuesCreated     <- FALSE
    maturitiesCreated <- FALSE

    for (i in 1:length(Symbols)) {
        if (verbose)
            cat("processing ", Symbols[[i]], ".....\n\n")
        default.return.class <- return.class

        if(str_detect(Symbols[[i]], "^TRISS")) {
          if(!issuesCreated) {
            formatReportAllotment(treasuries, "ISS", "issue_date", "maturity_date") %>%
              as.data.table(na.rm= FALSE) %>%
                setkey(security) -> issues
            issuesCreated <- TRUE
            }
          treasuries_i <- issues
        } else
        if(str_detect(Symbols[[i]], "^TRMTR")) {
          if(!maturitiesCreated) {
            formatReportAllotment(treasuries, "MTR", "maturity_date", "issue_date") %>%
              as.data.table(na.rm= FALSE) %>%
                setkey(security) -> maturities
            maturitiesCreated <- TRUE
            }
          treasuries_i <- maturities
        }
        else { stop(str_c("getSymbols.treasury has an invalid Symbol: ", Symbols[[i]])) }
        filter(treasuries_i, security == Symbols[[i]]) %>%  select(- "security") -> fr
        colnames(fr) <- str_c(Symbols[[i]], ".", colnames(fr))
        fr <- xts(as.matrix(fr[, -1]), as.Date(fr[, 1], origin = "1970-01-01"),
            src = "treasury", updated = Sys.time())
        colnames(fr) <- toupper(colnames(fr))
        fr <- quantmod___convert.time.series(fr = fr, return.class = return.class)
        if (auto.assign)
            assign(Symbols[[i]], fr, env)
    }
    if (auto.assign)
        return(Symbols)
    return(fr)
})}




